File: lab2/gr9sub2/Лапатецкий_Сергей/src/fork_example.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/wait.h>

int main(void)
{
	fflush(stdout);

	printf("parent(PID=%d)\n", getpid());

	pid_t pid[2];

	for(int i = 0; i < 2; ++i)
	{

		pid[i] = fork();

		switch(pid[i])
		{
			case -1:
				printf("Error\n");
			break;

			case 0:
				printf("%cchild[%d](PID=%d, PPID=%d)\n", '-', i, getpid(), getppid());
				sleep(5);
				exit(EXIT_SUCCESS);
			break;
		}
	}

	for(int i = 0; i < 2; ++i)
		wait(NULL);

	return 0;
}

File: lab2/src/fork_example.py
import os
import sys
import time

def terminate_process(process_name):
    print(f"{process_name}: PID = {os.getpid()}, PPID = {os.getppid()}", flush=True)
    sys.exit(0)

def execute_parent_and_children():
    print(f"Parent: My PID is {os.getpid()}", flush=True)
    first_child_pid = -1

    first_child_pid = os.fork()

    if first_child_pid == 0:
        terminate_process("Child A")

    if first_child_pid > 0:
        second_child_pid = os.fork()
        if second_child_pid == 0:
            terminate_process("Child B")

    time.sleep(50)

    if first_child_pid > 0 and second_child_pid > 0:
        child_1_pid, child_1_status = os.waitpid(first_child_pid, 0)
        child_1_exit_code = os.WEXITSTATUS(child_1_status)
        print(f"Parent: Child A (PID={child_1_pid}) finished with status {child_1_exit_code}", flush=True)

        child_2_pid, child_2_status = os.waitpid(second_child_pid, 0)
        child_2_exit_code = os.WEXITSTATUS(child_2_status)
        print(f"Parent: Child B (PID={child_2_pid}) finished with status {child_2_exit_code}", flush=True)

        print("Parent: All children have finished", flush=True)

if __name__ == "__main__":
    execute_parent_and_children()
